### ISSUE TITLE
Named arguments (function -a) are always set

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#2645]
When working on __fish_git_prompt, I wanted to use function's "-a"/"--argument-names" option, but I noticed that I then can't actually check if an argument has been given.

See:

``` fish
function t --argument a b
    echo $a $b
    set -q b; and echo "b is set: $b"; or echo "b is unset"
    set -q argv[2]; and echo "argv[2] is set: $argv[2]"; or echo "argv[2] is unset"
end
t foo
t foo bar
```

The output is:

> foo
> b is set: 
> argv[2] is unset
> 
> foo bar
> b is set: bar
> argv[2] is set: bar

**Comment by derekstavis** [comment]
I have noticed the same issue. Since then I have been using `test -z`:

```fish
function t --argument a b
    echo $a $b
    test -z b; and echo "b is set: $b"; or echo "b is unset"
    test -z argv[2]; and echo "argv[2] is set: $argv[2]"; or echo "argv[2] is unset"
end
t foo
t foo bar
```

The output is:

>foo
>b is unset: 
>argv[2] is unset
>
>foo bar
>b is unset
>argv[2] is unset

Anyway, I think that it makes a lot more sense to use `set -q`.

**Comment by krader1961** [comment]
If you add `echo num elements (count $a)` you'll see it has zero elements. So it's acting like this Python pattern `def t(a=()):` to assign an empty tuple to the var if the caller doesn't provide a value. This is going to need careful thought and discussion because any change is likely to break backward compatibility and we only want to do that once.

Q1: The existing semantics make your example the equivalent of inserting this block of code at the top of the function:

```
set -l a $argv[1]
set -l b $argv[2]
```

Which is why the named vars are considered to be defined. Should the semantics instead be changed to:

```
set -q argv[1]; and set -l a $argv[1]
set -q argv[2]; and set -l b $argv[2]
```

Note that if we only do this change and nothing else it shouldn't break anything since undefined and empty (i.e., zero elements) vars both expand to nothing. Or is there an edge case I haven't considered?

Q2: Should the presence of __--arguments__ imply that the arguments are mandatory or optional? The current behavior is optional. An argument could be made that the syntax implies they are mandatory.

Q3: Should the __function__ command be modified to make it possible to distinguish between mandatory and optional arguments?

**Comment by faho** [comment]
>set -q argv[1]; and set -l a $argv[1]

That's what I'm asking for, yes.

>it shouldn't break anything since undefined and empty (i.e., zero elements) vars both expand to nothing. Or is there an edge case I haven't considered?

The only thing that'll change is `set -q var`, which until this is implemented will always succeed for the named arguments, so it shouldn't actually be in any code.

>Q2: Should the presence of --arguments imply that the arguments are mandatory or optional?

There's some talk about this in the docopt issue (#478). With that you could define what is mandatory and what is optional.

With this implemented, it would become possible to check whether an argument has actually been given if it is allowed to be empty (i.e. there'd be a difference between `foo bar` and `foo bar ""`), so the function could implement the check itself even in the case of empty arguments.

**Comment by krader1961** [comment]
Okay, seeing that others have much grander ideas for expanding the function syntax let's keep this issue focused on just making the handling of unset __--argument__  values saner by changing the semantics to (in pseudo-code):

```
set -q argv[$n]; and set -l $argument[$n] $argv[$n]
```

This will affect existing code that tests if a var is defined when the function is invoked without a value for the var but technically such code is already semi-broken. Too, if there is code explicitly doing `set -q named_arg` like you tried presumably it will correctly handle the fact that all of a sudden the named_arg is no longer defined in this scenario.

If you haven't already started working on a fix, @faho, I'll take a stab at it.

**Comment by faho** [comment]
>If you haven't already started working on a fix, @faho, I'll take a stab at it.

I'm not working on it - my C++fu isn't all that great.

**Comment by gotjoshua** [comment]
is this still in progress one year and one month later?

**Comment by krader1961** [comment]
@gotjoshua: Yes, it is. Notice that there are over 500 open issues and just a handful of people creating patches on a regular basis (I'm one of them) and we're all doing it in our free time. Feel free to take a stab at fixing this or any other issue. Patches are always welcome. We'd love to see more people contributing changes other than command completions.

**Comment by krader1961** [comment]
This is another issue where changing the behavior is not backward compatible. So retargeting to next-major.

**Comment by gotjoshua** [comment]
Hi again,
Just a progress check and another 2 cents...
How is this going? Is there a branch started or could you point me at the part of the code that I could "take a stab" at?

Also, I wonder if there is a reason why fish does not follow the quite common and understandable syntax with arguments and defaults:
`function foo (reqarg1,optarg2=default)
`

was this a conscious decision to avoid such syntax?
is this discussion ongoing elsewhere?

**Comment by faho** [comment]
>Is there a branch started

Nope.

>could you point me at the part of the code that I could "take a stab" at?

It's actually apparently almost trivial - in `function_prepare_environment` (in src/function.cpp), there is

```c++
            if (*arg) {
                env_set_one(named_arguments.at(i), ENV_LOCAL | ENV_USER, *arg);
                arg++;
            } else {
                env_set_empty(named_arguments.at(i), ENV_LOCAL | ENV_USER);
            }
```

What needs to be done there to fulfill my original request is simply to remove the `else` branch or replace it with a `break`. Then this needs a whole bunch of testing.

However, I have since discovered that `set -q var[1]` will work, so I'm not really all that sure that this is necessary.

>was this a conscious decision to avoid such syntax?

Well, that particular syntax won't really work since `()` is fish's syntax for command substitution (i.e. it would try to execute it as a command).

As for why that particular syntax was picked... I don't know. It was added in 2007 by the original maintainer, who has since left the project.

> is this discussion ongoing elsewhere?

You mean discussion on improving named arguments? Adding required ones and so on?

I'm not aware. However, that would make a nice new issue, so if you're interested, think a bit about what specifically you would like and open a new issue with a proposal.
